### PR TITLE
Add payment

### PR DESCRIPTION
**Pull request by fadyyousry**
- Add gem 'stripe' payment (apply version 3)
- Add gem 'stripe_event' to handle callbacks (event)
- add a view for in-cart items
- table for in-cart items
    * user_id
    * book_id
    * timestamp
    * status [in progress, completed]
- add to cart --> sign in (if not signed in)
- add to cart button not appear for admin
- update ability
- user can't buy a book twice

### CODE REVIEW

File: app/controllers/bills_controller.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+class BillsController < ApplicationController
+
+  def new
+    set_book
+    set_product
```
Comment (Magdyx) [inline app/controllers/bills_controller.rb:5]: move to model

File: app/services/stripe/payment_intent_event_handler.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+            end
+        end
+
+        def handle_payment_intent_payment_failed(event)
+
+        end
+
+        def handle_payment_intent_succeeded(event)
+            
+        end
```
Comment (Magdyx) [inline app/services/stripe/payment_intent_event_handler.rb:20]: what are those for?

File: app/controllers/bills_controller.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
+class BillsController < ApplicationController
```
Comment (Magdyx) [inline app/controllers/bills_controller.rb:1]: no ability checking?

File: app/controllers/bills_controller.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      @book = Book.find(params[:book])
+    end
+
+    def set_product
+      if @book.product_id.nil?
+        @product = Stripe::Product.create({
+                    name: @book.title,
+                    images: [@book.image_link],
+                    description: "Description: " + @book.description
+                  })
```
Comment (Magdyx) [inline app/controllers/bills_controller.rb:36]: Is there any kind of exceptions thrown from the create method that we should handle?

File: app/services/stripe/payment_intent_event_handler.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+module Stripe
+    class PaymentIntentEventHandler
+        def call(event)
+            begin
+                method = "handle_" + event.type.tr('.', '_')
+                self.send method, event
```
Comment (Magdyx) [inline app/services/stripe/payment_intent_event_handler.rb:6]: can we remove self here?

File: app/views/bills/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+<%= link_to t('defualt.buy_new'), new_bill_path(book: book), remote: true,
```
Comment (Magdyx) [inline app/views/bills/_form.html.erb:1]: typo

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -12,6 +12,8 @@
 
     <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
     <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_include_tag 'https://js.stripe.com/v3/', 'data-turbolinks-track': 'reload' %>
+
```
Comment (Magdyx) [inline app/views/layouts/application.html.erb:16]: can this be moved to the application.js file somehow?

File: app/controllers/bills_controller.rb (hunk 3 above)
Comment (fadyyousry) [reply app/controllers/bills_controller.rb:1]: how can I check the ability for controller only not (model)?

File: app/services/stripe/payment_intent_event_handler.rb (hunk 5 above)
Comment (fadyyousry) [reply app/services/stripe/payment_intent_event_handler.rb:6]: ok, no problem

File: app/views/layouts/application.html.erb (hunk 7 above)
Comment (fadyyousry) [reply app/views/layouts/application.html.erb:16]: I don't know

File: app/assets/stylesheets/template/total_amount.css [hunk 8]
```diff
@@ -0,0 +1,68 @@
+.btn-widget {
```
Comment (Magdyx) [inline app/assets/stylesheets/template/total_amount.css:1]: **Comment on the file:** 
1. file location shouldn't be in template: only template stylesheets are kept their.
2. please use sass extension `.scss` in all new stylesheet files

File: app/models/sale.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
+class Sale < ApplicationRecord
+  paginates_per 12
+  belongs_to :user, required: true
+  belongs_to :book, required: true
+
+  attribute :status, default: 'In Progress', presence: true
```
Comment (Magdyx) [inline app/models/sale.rb:6]: you could have used enum for status and save it as integer not string

File: app/models/user.rb [hunk 10]
```diff
@@ -3,14 +3,24 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
-
+  has_many :sales
+  has_many :books, through: :sales
```
Comment (Magdyx) [inline app/models/user.rb:7]: usually add `dependent: do something` in the association part

File: app/models/user.rb [hunk 11]
```diff
@@ -3,14 +3,24 @@ class User < ApplicationRecord
...
+  
+  enum status: [ :in_progress, :completed ]
```
Comment (Magdyx) [inline app/models/user.rb:9]: this line is confusing. why the user has status

File: app/views/books/show.html.erb [hunk 12]
```diff
@@ -72,10 +73,16 @@
...
+                <% unless current_user_admin? %>
                 <div class="btn-sec">
-                    <button class="btn "><%= t('default.add_to_cart')%></button>
-                    <button class="btn black"><%= t('defualt.buy_new')%></button>
+                    <% if current_user_customer? %>
+                    <%= link_to t('default.add_to_cart'), sales_path(book_id: @book.id),
+                        method: :post, class: "btn" %>
+                    <% else %>
+                    <%= link_to t('default.add_to_cart'), new_user_session_path, class: "btn" %>
+                    <% end %>
```
Comment (Magdyx) [inline app/views/books/show.html.erb:83]: fix indentation

File: app/views/books/show.html.erb [hunk 13]
```diff
@@ -72,10 +73,16 @@
                         <span class="value final">$<%= @book.price %></span>
                     </li>
                 </ul>
+                <% unless current_user_admin? %>
                 <div class="btn-sec">
-                    <button class="btn "><%= t('default.add_to_cart')%></button>
-                    <button class="btn black"><%= t('defualt.buy_new')%></button>
+                    <% if current_user_customer? %>
```
Comment (Magdyx) [inline app/views/books/show.html.erb:78]: I think you missed can't buy book twice part.

File: app/views/shared/_books.html.erb [hunk 14]
```diff
@@ -8,7 +8,7 @@
       <% books.each do |book| %>
         <div class="col-lg-3 col-md-6 mb-4">
           <div class="item">
-            <%= image_tag book.image_link, size: Book::IMAGE_SIZE , alt: "img" %>
+            <%= image_tag book.image_link, size: "141x218", alt: "img" %>
```
Comment (Magdyx) [inline app/views/shared/_books.html.erb:11]: why are you using hardcoded value?
Also can you remove the alt

File: app/views/shared/_navbar.html.erb [hunk 15]
```diff
@@ -36,11 +36,17 @@
...
-        <i class="fa fa-shopping-cart" aria-hidden="true"></i></span>
-      <span class="quntity">3</span>
-    </div>
+    <% if current_user_customer? %>
+      <div class="cart my-2 my-lg-0">
+        <%= link_to sales_path, class: "text-warning" do %>
+          <span>
+            <i class="fa fa-shopping-cart" aria-hidden="true"></i></span>
+            <% if cart_size > 0 %>
+              <span class="quntity"><%= cart_size %></span>
```
Comment (Magdyx) [inline app/views/shared/_navbar.html.erb:45]: typo in class

File: app/views/shared/_navbar.html.erb [hunk 16]
```diff
@@ -36,11 +36,17 @@
...
-      <span>
-        <i class="fa fa-shopping-cart" aria-hidden="true"></i></span>
-      <span class="quntity">3</span>
-    </div>
+    <% if current_user_customer? %>
+      <div class="cart my-2 my-lg-0">
+        <%= link_to sales_path, class: "text-warning" do %>
+          <span>
+            <i class="fa fa-shopping-cart" aria-hidden="true"></i></span>
+            <% if cart_size > 0 %>
```
Comment (Magdyx) [inline app/views/shared/_navbar.html.erb:44]: do we really need this if?

File: app/services/stripe/payment_intent_event_handler.rb [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+            end
+        end
+
+        def handle_payment_intent_payment_failed(event)
+
+        end
+
+        def handle_payment_intent_succeeded(event)
+            user = User.find_by(customer_id: event[:data][:object][:customer])
+            user.sales.where(status: 'In Progress').update_all(status: 'Completed')
```
Comment (Magdyx) [inline app/services/stripe/payment_intent_event_handler.rb:20]: I think you should handle user is null case here

File: app/controllers/bills_controller.rb [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+        payment_method_types: ['card'],
+        customer: current_user.customer_id,
+        line_items: get_line_items,
+        mode: 'payment',
+        success_url: books_url,
+        cancel_url: sales_url,
+      })
+    rescue => e
+      Rails.logger.debug e.message
+    end
```
Comment (Magdyx) [inline app/controllers/bills_controller.rb:15]: just log the exception? nothing to notify the user?

File: app/controllers/bills_controller.rb [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+
+  private
+    def sales_ids
+      params.require(:sales)
+    end
+
+    def get_books
+      books = []
+      sales_ids.each do |id|
+        book = Book.find(Sale.find(id).book_id)
```
Comment (Magdyx) [inline app/controllers/bills_controller.rb:29]: find throws an exception if item not found, you should handle that in your logic

File: app/controllers/bills_controller.rb [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    def get_line_items
+      line_items = []
+      get_books.each do |book|
+        item = Hash.new
+        item[:price_data] = Hash.new
+        item[:price_data][:product] = book.product_id
+        item[:price_data][:unit_amount] = (book.price * 100).to_i
+        item[:price_data][:currency] = 'usd'
+        item[:quantity] = 1
+        line_items.append(item)
```
Comment (Magdyx) [inline app/controllers/bills_controller.rb:44]: I think this part is better be done as a separate method.

File: app/controllers/bills_controller.rb [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+      books
+    end
+
+    def get_line_items
+      line_items = []
+      get_books.each do |book|
+        item = Hash.new
+        item[:price_data] = Hash.new
+        item[:price_data][:product] = book.product_id
+        item[:price_data][:unit_amount] = (book.price * 100).to_i
```
Comment (Magdyx) [inline app/controllers/bills_controller.rb:41]: this operation is vague, please move it to the book model

File: app/controllers/sales_controller.rb [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    layout 'application'
+    load_and_authorize_resource
+
+    def index
+      @sales = @sales.where(status: "In Progress").page(params[:page])
+    end
+    
+    def create
+      @sale = Sale.new (sale_params)
+      @sale.user_id = current_user.id
```
Comment (Magdyx) [inline app/controllers/sales_controller.rb:11]: these 2 lines can be replaced by
```
@sale = current_user.sales.new(sale_params)
```

File: app/controllers/sales_controller.rb [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+      @sale.save
+      if @sale.save
```
Comment (Magdyx) [inline app/controllers/sales_controller.rb:13]: whoa? you meant to do that?

File: app/helpers/sale_helper.rb [hunk 24]
```diff
@@ -0,0 +1,9 @@
+module SaleHelper
+    def total_price
+        price = 0
+        @sales.each do |sale|
+            price += sale.book.price
+        end
+        price
+    end
```
Comment (Magdyx) [inline app/helpers/sale_helper.rb:8]: I think there is a better way to sum array of elements
maybe try [this](https://apidock.com/rails/Enumerable/sum)

File: app/models/book.rb [hunk 25]
```diff
@@ -73,4 +78,13 @@ def cascade_categories
         end
       end
     end
+
+    def create_product
+      product = Stripe::Product.create({
+                  name: title,
+                  images: [image_link],
+                  description: description
+                })
```
Comment (Magdyx) [inline app/models/book.rb:87]: handle stripe exceptions

File: app/views/layouts/application.html.erb (hunk 7 above)
Comment (Magdyx) [reply app/views/layouts/application.html.erb:16]: I thought you can use `yarn add` for that

File: db/migrate/20200719213329_create_sales.rb [hunk 26]
```diff
@@ -0,0 +1,12 @@
+class CreateSales < ActiveRecord::Migration[6.0]
+  def change
+    create_table :sales do |t|
+      t.references :user, null: false, foreign_key: true
+      t.references :book, null: false, foreign_key: true
```
Comment (Magdyx) [inline db/migrate/20200719213329_create_sales.rb:5]: what happens if you delete a user or a book

File: app/controllers/bills_controller.rb [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+      redirect_to sales_url, notice: t('default.messages.no_item')
+    end
+    respond_to do |format|
+      format.js
+    end
+  end
+  
+  private
+    def products_params
+      params.require(:products).map {|product| product.permit(:product_id, :price)}
```
Comment (Magdyx) [inline app/controllers/bills_controller.rb:25]: sorry. what is the structure of the params

File: app/controllers/bills_controller.rb [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+    end
+    
+    def authorize_checkout
+      authorize! :create, Sale
+    end
+    
+    def get_line_items
+      products_params.map do |product|
+        create_item product
+        end
```
Comment (Magdyx) [inline app/controllers/bills_controller.rb:35]: fix indentation

File: app/controllers/manager/books_controller.rb [hunk 29]
```diff
@@ -36,8 +41,12 @@ def update
     end
   
     def destroy
-      @book.destroy
-      redirect_to manager_books_url, notice: t("messages.success.destroy", model: @book.class.name)
+      begin
+        @book.destroy
+        redirect_to manager_books_url, notice: t("messages.success.destroy", model: @book.class.name)
+      rescue
```
Comment (Magdyx) [inline app/controllers/manager/books_controller.rb:47]: what exception are you catching here.
please use Rails Logger to log exceptions as well

File: app/controllers/manager/users_controller.rb [hunk 30]
```diff
@@ -24,6 +29,7 @@ def update
 
     def destroy
       @user.destroy!
+      redirect_to manager_users_url, notice: t("messages.success.destroy", model: @user.class.name)
```
Comment (Magdyx) [inline app/controllers/manager/users_controller.rb:32]: good catch

File: app/controllers/manager/users_controller.rb [hunk 31]
```diff
@@ -24,6 +29,7 @@ def update
 
     def destroy
       @user.destroy!
```
Comment (Magdyx) [inline app/controllers/manager/users_controller.rb:31]: I think there is an exception I should have caught here

File: app/helpers/sale_helper.rb [hunk 32]
```diff
@@ -0,0 +1,11 @@
+module SaleHelper
+    def total_price
+        @sales.sum do |sale|
+            sale.book.price
+        end
+    end
+
+    def extract_products_params
+        (@sales.map {|sale| sale.book}).as_json(only: [:product_id, :price])
+    end
```
Comment (Magdyx) [inline app/helpers/sale_helper.rb:10]: what are you using as_json for?

File: app/controllers/manager/users_controller.rb (hunk 31 above)
Comment (fadyyousry) [reply app/controllers/manager/users_controller.rb:31]: no user's dependence is nullify so no need to catch something here

File: app/helpers/sale_helper.rb (hunk 32 above)
Comment (fadyyousry) [reply app/helpers/sale_helper.rb:10]: to make it return hash not array

File: app/views/layouts/application.html.erb (hunk 7 above)
Comment (fadyyousry) [reply app/views/layouts/application.html.erb:16]: it didn't work

File: app/controllers/bills_controller.rb (hunk 28 above)
Comment (fadyyousry) [reply app/controllers/bills_controller.rb:35]: ?

File: app/controllers/bills_controller.rb (hunk 27 above)
Comment (fadyyousry) [reply app/controllers/bills_controller.rb:25]: `<ActionController::Parameters {"products"=>[{"price"=>"10.0", "product_id"=>"prod_HgKvRc1srJCgWA"}, {"price"=>"15.0", "product_id"=>"prod_HgKuW3VlYPZxpz"}], "controller"=>"bills", "action"=>"new"} permitted: false>`

**Pull request merged**